### ISSUE TITLE
Argument Count Error

### ISSUE THREAD

**Issue by wessleym** [google/closure-compiler#730]
I am running the version of Closure Compiler that I downloaded yesterday.
I would think the below code would produce a warning regarding the number of arguments used in the call to `staticMethod` (within `onload`).  It does not.  But if `privatelyUsedMethod`'s declaration and reference (within `staticMethod`) are deleted, I get the warning I would expect.

``` js
var StaticClass = {
  privatelyUsedMethod : function() {},
  staticMethod : function(msg) {
    alert(msg);
    StaticClass.privatelyUsedMethod();
  }
};
window.onload = function() { StaticClass.staticMethod(); };
```

Also, I would think the below code would give me a warning too, but since `StaticClass` is declared after the reference to `staticMethod` (in `onload`), no warnings are displayed.  But when I reference `staticMethod` after declaring `StaticClass`, I see the warnings I would expect.

``` js
window.onload = function() { StaticClass.staticMethod(); };
var StaticClass = {
  staticMethod : function(msg) {
    alert(msg);
  }
};
```

I believe these are bugs.  Would someone help please?  Thanks.

**Comment by wessleym** [comment]
@concavelenz, thank you.  Shouldn't both circumstances produce warnings?  I would think so.  I want to make sure I'm understanding you.

**Comment by concavelenz** [comment]
It should but I've seen problems around this before.  Hopefully, the new type inference handles circular type references correctly.

**Comment by wessleym** [comment]
@dimvar and @concavelenz, I've been checking back on this issue for years, watching for any activity.  Just for the record, this bug is still present.  Thank you for your time.

**Comment by rishipal** [comment]
Created internal issue - http://b/140445611

**Comment by shicks** [comment]
I suspect that the issue here is that `StaticClass` is not treated as a namespace, so the compiler does not keep track of the types of its properties as strictly.  It's instead just an "ad-hoc" object type, so I think the function types are treated as "inferred" rather than "declared", making it more permissive.  You can see the issue [in the debugger][var] by looking at the bottom of the generate AST - the final GETPROP node (third line from the bottom) has type `?` instead of `function(?): undefined`.

If you mark it as `@const` (or use the `const` keyword) then you [do get the errors you're looking for][const].

So the question is what makes the difference?  It looks like it's [the recursive reference to `StaticClass`][rec], though it's also seemingly important that the call be in a closure, rather than top-level (removing the `onload` also recovers the expected error).

I don't have a clear picture of the interactions here, but it seems like aggressive use of `const` should address any issues - given that this is an easy workaround, is this actually blocking anything?

[var]: https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520StaticClass%2520%253D%2520%257B%250A%2520%2520privatelyUsedMethod%2520%253A%2520function()%2520%257B%257D%252C%250A%2520%2520staticMethod%2520%253A%2520function(msg)%2520%257B%250A%2520%2520%2520%2520alert(msg)%253B%250A%2520%2520%2520%2520StaticClass.privatelyUsedMethod()%253B%250A%2520%2520%257D%250A%257D%253B%250Awindow.onload%2520%253D%2520function()%2520%257B%2520StaticClass.staticMethod()%253B%2520%257D%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

[const]: https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540const%2520*%252F%250Avar%2520StaticClass%2520%253D%2520%257B%250A%2520%2520privatelyUsedMethod%2520%253A%2520function()%2520%257B%257D%252C%250A%2520%2520staticMethod%2520%253A%2520function(msg)%2520%257B%250A%2520%2520%2520%2520alert(msg)%253B%250A%2520%2520%2520%2520StaticClass.privatelyUsedMethod()%253B%250A%2520%2520%257D%250A%257D%253B%250Awindow.onload%2520%253D%2520function()%2520%257B%2520StaticClass.staticMethod()%253B%2520%257D%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

[rec]: https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520StaticClass%2520%253D%2520%257B%250A%2520%2520staticMethod%2520%253A%2520function(msg)%2520%257B%250A%2520%2520%2520%2520alert(StaticClass)%253B%250A%2520%2520%257D%250A%257D%253B%250Awindow.onload%2520%253D%2520function()%2520%257B%2520StaticClass.staticMethod()%253B%2520%257D%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

**Comment by wessleym** [comment]
Thank you for the suggestion.  I can add `const`.  I still see this as a bug since I might easily forget to add `const` to one of my static classes.  But at least this is a workaround.  Thank you again.
